core/gallery/cache.go

```diff
@@ -0,0 +1,231 @@
+package gallery
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/anyproto/any-sync/app"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/anyproto/anytype-heart/core/wallet"
+	"github.com/anyproto/anytype-heart/pb"
+	"github.com/anyproto/anytype-heart/pkg/lib/pb/model"
+)
+
+const (
+	CacheCName = "gallery-index-cache"
+
+	galleryDir    = "gallery"
+	indexFileName = "index.pb"
+	verFileName   = "ver"
+
+	indexURI = "https://tools.gallery.any.coop/app-index.json"
+)
+
+type IndexCache interface {
+	app.Component
+
+	GetManifest(downloadLink string, timeoutInSeconds int) (info *model.ManifestInfo, err error)
+	GetIndex(timeoutInSeconds int) (*pb.RpcGalleryDownloadIndexResponse, error)
+}
+
+type cache struct {
+	storage  cacheStorage
+	indexURL string
+}
+
+func NewCache() IndexCache {
+	return &cache{}
+}
+
+func (c *cache) Init(a *app.App) error {
+	path := filepath.Join(app.MustComponent[wallet.Wallet](a).RepoPath(), galleryDir)
+	if err := os.Mkdir(path, 0777); err != nil && !os.IsExist(err) {
+		return fmt.Errorf("failed to init gallery index directory: %w", err)
+	}
+
+	c.storage = &storage{
+		versionPath: filepath.Join(path, verFileName),
+		indexPath:   filepath.Join(path, indexFileName),
+	}
+
+	c.indexURL = indexURI
+	return nil
+}
+
+func (c *cache) Name() string {
+	return CacheCName
+}
+
+func (c *cache) GetIndex(timeoutInSeconds int) (*pb.RpcGalleryDownloadIndexResponse, error) {
+	localIndex, err := c.storage.getIndex()
+	if err != nil {
+		log.Warn("failed to read local index. Need to refetch index from remote", zap.Error(err))
+	}
+
+	version := ""
+	if localIndex != nil {
+		version, err = c.storage.getVersion()
+		if err != nil {
+			log.Warn("failed to read local version. Need to refetch version from remote", zap.Error(err))
+		}
+	}
+
+	index, newVersion, err := c.downloadGalleryIndex(timeoutInSeconds, version, false)
+	if err != nil {
+		if errors.Is(err, ErrNotModified) {
+			return localIndex, nil
+		}
+
+		if localIndex != nil {
+			log.Warn("failed to download index from remote. Returning local index", zap.Error(err))
+			return localIndex, nil
+		}
+
+		return nil, err
+	}
+
+	c.storage.save(index, newVersion)
+
+	return index, nil
+}
+
+func (c *cache) GetManifest(downloadLink string, timeoutInSeconds int) (info *model.ManifestInfo, err error) {
+	localIndex, err := c.storage.getIndex()
```

Comment:
May be extract some duplicate code from GetIndex and GetManifest? Looks like they have a lot of common code?

Reply:
Yes, it hepled to remove 30 lines :smile: